django/thunderstore/core/storage.py

```diff
@@ -46,11 +47,20 @@ def save(
         Calling .save() closes the file, so use temporary copies for
         mirrors and call the main bucket with the actual file last.
         """
-        for storage_mirror in self.mirrors:
-            with TemporarySpooledCopy(content) as tmp_content:
-                storage_mirror.save(name, tmp_content, max_length)
 
-        return super().save(name, content, max_length)
+        cache_key = f"cache_mirror_storage_{name}"
+
+        if not cache.add(cache_key, "LOCKED", timeout=5):
+            raise Exception("Another save operation is in progress for this file.")
+
+        try:
+            for storage_mirror in self.mirrors:
+                with TemporarySpooledCopy(content) as tmp_content:
+                    storage_mirror.save(name, tmp_content, max_length)
+
+            return super().save(name, content, max_length)
+        finally:
+            cache.delete(cache_key)
```

Comment:
Don't use Django cache directly, Redis has first party support for locking which should be used. Check `thunderstore.cache.utils.get_cache` and how it's utilized for locking in `thunderstore.cache.cache.try_regenerate_cache` for example